mongo.go

```diff
@@ -0,0 +1,472 @@
+	}
+	ctx, cls := context.WithTimeout(context.Background(), m.opTimeout)
+	defer cls()
+	return m.get(ctx, key)
+}
+
+func (m *MongoDS) Delete(key datastore.Key) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	if m.closed {
+		return ErrClosed
+	}
+
+	ctx, cls := context.WithTimeout(context.Background(), m.opTimeout)
+	defer cls()
+	return m.delete(ctx, key)
+}
+
+func (m *MongoDS) QueryExtended(q dsextensions.QueryExt) (query.Results, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	if m.closed {
+		return nil, ErrClosed
+	}
+
+	ctx, cls := context.WithTimeout(context.Background(), m.opTimeout)
+	defer cls()
+
+	return m.query(ctx, q)
+}
+
+func (m *MongoDS) Query(q query.Query) (query.Results, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	if m.closed {
+		return nil, ErrClosed
+	}
+
+	ctx, cls := context.WithTimeout(context.Background(), m.opTimeout)
+	defer cls()
+
+	qe := dsextensions.QueryExt{Query: q}
+
+	return m.query(ctx, qe)
+}
+
+func (m *MongoDS) Close() error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.closed {
+		return fmt.Errorf("mongods already closed")
+	}
+	ctx, cls := context.WithTimeout(context.Background(), m.opTimeout)
+	defer cls()
+	if err := m.m.Disconnect(ctx); err != nil {
+		return fmt.Errorf("client disconnecting: %s", err)
+	}
+	m.closed = true
+	return nil
+}
+
+func (m *MongoDS) get(ctx context.Context, key datastore.Key) ([]byte, error) {
+	sr := m.col.FindOne(ctx, bson.M{"_id": key.String()})
+	if sr.Err() == mongo.ErrNoDocuments {
+		return nil, datastore.ErrNotFound
+	}
+	if sr.Err() != nil {
+		return nil, fmt.Errorf("delete document: %s", sr.Err())
+	}
+	var kv keyValue
+	if err := sr.Decode(&kv); err != nil {
+		return nil, fmt.Errorf("decoding key-value: %s", err)
+	}
+	return kv.Value, nil
+}
+
+func (m *MongoDS) delete(ctx context.Context, key datastore.Key) error {
+	_, err := m.col.DeleteOne(ctx, bson.M{"_id": key.String()})
+	if err != nil {
+		return fmt.Errorf("delete document: %s", err)
+	}
+
+	return nil
+}
+
+func (m *MongoDS) put(ctx context.Context, key datastore.Key, val []byte) error {
+	_, err := m.col.UpdateOne(ctx, bson.M{"_id": key.String()}, bson.M{"$set": bson.M{"v": val}}, options.Update().SetUpsert(true))
+	if err != nil {
+		return fmt.Errorf("inserting/updating key-value: %s", err)
+	}
+	return nil
+}
+
+func (m *MongoDS) has(ctx context.Context, key datastore.Key) (bool, error) {
+	sr := m.col.FindOne(ctx, bson.M{"_id": key.String()})
+	if sr.Err() == mongo.ErrNoDocuments {
+		return false, nil
+	}
+	if sr.Err() != nil {
+		return false, fmt.Errorf("finding key: %s", sr.Err())
+	}
+	return true, nil
+}
+
+func (m *MongoDS) getSize(ctx context.Context, key datastore.Key) (int, error) {
+	v, err := m.get(ctx, key)
+	if err == datastore.ErrNotFound {
+		return -1, err
+	}
+	if err != nil {
+		return 0, fmt.Errorf("getting value: %s", err)
+	}
+	return len(v), nil
+}
+
+func (m *MongoDS) query(ctx context.Context, q dsextensions.QueryExt) (query.Results, error) {
```

Comment:
So this setup is based on most implementations with some tunning around mongo way of doing things and cases optimizations.